tools/vswhere/MSBuildLocator.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Xamarin.Android.VSWhere
+{
+	/// <summary>
+	/// https://github.com/Microsoft/vswhere/wiki/Find-MSBuild
+	/// </summary>
+	public class MSBuildLocator
+	{
+		/// <summary>
+		/// Adds the -prerelease flag, defaults to true
+		/// </summary>
+		public bool IncludePrerelease { get; set; } = true;
+
+		public string VisualStudioDirectory { get; private set; }
+
+		public string MSBuildPath { get; private set; }
+
+		public void Locate ()
+		{
+			var programFiles = Environment.GetFolderPath (Environment.SpecialFolder.ProgramFilesX86);
+			var vswhere = Path.Combine (programFiles, "Microsoft Visual Studio", "Installer", "vswhere.exe");
+			if (!File.Exists (vswhere))
+				throw new FileNotFoundException ("Cannot find vswhere.exe!", vswhere);
+
+			string prerelease = IncludePrerelease ? "-prerelease" : "";
+			VisualStudioDirectory = Exec (vswhere, $"-latest {prerelease} -products * -requires Microsoft.Component.MSBuild -property installationPath");
+
+			// Try VS 2019 & Higher
+			MSBuildPath = Path.Combine (VisualStudioDirectory, "MSBuild", "Current", "Bin", "MSBuild.exe");
+			if (!File.Exists (MSBuildPath)) {
```

Comment:
Instead of this style of construct -- which means possible endless nesting as new versions are supported -- let's instead use an early exit:

```csharp
if (File.Exists (MSBuildPath))
    return;

// VS 2017
MSBuildPath = Path.Combine (VisualStudioDirectory, "MSBuild", "15.0", "Bin", "MSBuild.exe");
if (File.Exists (MSBuildPath))
    return;

// VS2015? ...
// ...

// bail; cannot find
throw new FileNotFoundException ("Cannot find MSBuild.exe!")
```

Reply:
We have also in the past (see xamarin-android-tools) just had an array of paths we check so we can easily just add a new one without having to add a ton of extra `Exists` stuff